fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/LoanApplicationTerms.java

```diff
@@ -228,6 +229,76 @@ public final class LoanApplicationTerms {
     private final boolean enableAccrualActivityPosting;
     private final List<LoanSupportedInterestRefundTypes> supportedInterestRefundTypes;
 
+    public static LoanApplicationTerms assembleFrom(LoanRepaymentScheduleModelData modelData) {
+        return new LoanApplicationTerms(modelData.currency(), // currency
```

Comment:
please check if you could use the Builder pattern instead here. This pattern looks horrible:
```
                null, // calendarHistoryDataWrapper
                false, // isInterestChargedFromDateSameAsDisbursalDateEnabled
                null, // numberOfDays
                false, // isSkipRepaymentOnFirstDayOfMonth
                null, // holidayDetailDTO
                false, // allowCompoundingOnEod
                false, // isEqualAmortization
                false, // false (last boolean value)
                false, // isInterestToBeRecoveredFirstWhenGreaterThanEMI
                null, // fixedPrincipalPercentagePerInstallment
                false, // isPrincipalCompoundingDisabledForOverdueLoans
                false, // enableDownPayment
                null, // disbursedAmountPercentageForDownPayment
                false, // isAutoRepaymentForDownPaymentEnabled
                null, // repaymentStartDateType
                null, // submittedOnDate
                null, // loanScheduleType
                null, // loanScheduleProcessingType
                modelData.fixedLength(), // fixedLength
                false, // enableAccrualActivityPosting
                List.of() // supportedInterestRefundTypes
````

Reply:
Using the builder pattern in this case is challenging due to the large number of final fields in the LoanApplicationTerms class. These fields must be initialized either directly in the constructor or through the builder, which requires adding them all to the constructor to satisfy the requirements of final fields.

As such, using a constructor to set all the fields (including null and default values) is necessary to avoid issues with initializing final fields that do not have default values. Otherwise, the builder cannot properly work without these mandatory parameters.

However, I'm open to discussing any ideas you might have on refactoring the model to make it more suitable for using the builder pattern.

What do you think?

Reply:
Why don't we simply drop the `final` modifier from those fields?

Reply:
Added a builder, but had to make it not an annotation, since it conflicts with constructors, and it is difficult to change old constructors to a builder, since they contain a lot of logic that pulls a chain of consequences.